### PR TITLE
[EventHubs] Track Two: Third Preview (Event Processor checkpointing and new public API)

### PR DESCRIPTION
**Pull request by kinelski**
These changes include:
- Support for checkpointing in Event Processor.
  - `EventProcessor` starts processing events from the last checkpointed sequence number if it exists.
  - `InMemoryPartitionManager`'s `UpdateCheckpointAsync` works properly. It updates the existing ownership with the provided `Checkpoint` information, updates its `LastModifiedTime` and its `ETag` as well.
  - Two live tests added to verify `EventProcessor`'s behavior.
  - Six tests added to verify `InMemoryPartitionManager`'s behavior.
- New public API for preview 3.
  - `IPartitionProcessor` interface replaced by `BasePartitionProcessor` class.
  - `EventProcessor` is a generic type, but it still supports a factory constructor.
  - `CheckpointManager` completely removed and its contents were merged into `PartitionContext`.
  - Changes in the SDK, tests, tests helpers and samples were made to keep things **almost*** working.
  - A lot of XML comments were updated as well.
- `InMemoryPartitionManager` takes consumer group and Event Hub name as key.
  - Three keys are now supported as a 3-element value tuple (partitionId, consumerGroup, eventHubName).
  - The two failing tests we had for `InMemoryPartitionManager` are now passing.

These changes do not include:
- Event Hubs namespace as a key for ownership in`InMemoryPartitionManager`. This will probably be added in this same PR (or in another one) during the course of the day.

*. Sample 12 was updated accordingly, but it needs some improvement to wait for load balancing to finish, otherwise it won't behave as expected.

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs]: We may want to consider tuning the language here a bit.  Maybe something like:

`Interacts with the storage system with responsibility for creation of checkpoints`

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs]: I'd suggest validating if there are any rules around these.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs]: What if offset or sequence number are `null`?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs]: I'd suggest just returning the task from `UpdateCheckpointAsync`;  you're not using it so there is no need to await.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs]: These methods should probably be `virtual` to allow for mocking.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionOwnership.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionOwnership.cs]: This is a bit odd.   Only created by a limited scope, but open for changing properties to `public`.    Was that intentional?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionOwnership.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionOwnership.cs]: I'd suggest `ArgumentAtLeast` here for the `Int64` members.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs]: nit:  Does `Associated` buy us anything here?   Are we taking a second instance somewhere that I overlooked where we could be confused by the scope?

**Review by jsquire — COMMENTED** [review]
Still working my way though the tests, but I wanted to release comments thus far.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/BasePartitionProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/BasePartitionProcessor.cs]: I believe one of the points in replacing the interface with the abstract class was to allow users to only implement the methods they needed, as things starts to get overwhelming at first. The methods were supposed to do some kind of logging, but we currently don't have that in Event Processor, so there's nothing to do.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs]: I don't know what's going on with my brain.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs]: I was taking the struct approach, but I found out about value tuples midway through. I'll decide on that in a few moments, but I'll probably revert it to a struct if people prefer that.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs]: The `Logger` is just a temporary thing we have while `EventProcessor` doesn't have an event source.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionOwnership.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionOwnership.cs]: Yes. The main idea was not letting the users instantiate `PartitionOwnership` (unless for mocking purposes) and making some information immutable (`PartitionId`, `EventHubName`, `ConsumerGroup` and `OwnerIdentifier`). However, some properties need to be changed (`Offset`, `SequenceNumber` and `LastModifiedTime` upon checkpointing, `ETag` upon claim). Do you think there's a better way of doing that?

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Processor/EventProcessorLiveTests.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Processor/EventProcessorLiveTests.cs]: Some of these scenarios feel like we should be able to mock them.    For example, I'm not sure why we would need to use Azure to make sure that the calling start calls the partition's initialize method.   That should be something we can do in isolation.

Not a priority for preview 3, but let's make sure that we have an item to revisit the tests and review them for preview 4.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs]: No. Just didn't want to keep the `PartitionContext`. Does `Context` seem reasonable?

**Review by jsquire — APPROVED** [review]
Nothing that I would consider blocking, but some things that we should probably consider before merging.  Marking as approved so that I don't block code complete.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/BasePartitionProcessor.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/BasePartitionProcessor.cs]: Fair enough.  The class should probably still be `abstract` and I'd advocate for the process and error methods as well.   I can see the case for Initialize and Stop to be no-ops.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs]: Your call.    I was just mentioning.   Unlikely that you'll attract attention of those who have strong opinions, and you have good justification with it being a `private` member.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs]: Ah, that makes sense, then.

Would you be so kind as to log an issue for creating the Event Source for this?   We should probably pull that into preview 4.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionOwnership.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionOwnership.cs]: Nope, I think you've got the right approach for your scenario.  I don't have a reason against.   It just looked a bit inconsistent and I wasn't sure if there was some refactoring that got forgotten.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs]: I like `Context`  👍

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/Checkpoint.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/Checkpoint.cs]: I think it's valid to say it's only used in the context of an event processor. We're not expecting the user to instantiate checkpoints (unless they create a class from it, but this is allowed only for mocking purposes). In our code, Checkpoints are only instantiated by the `PartitionContext` when requested by the user, and we have no intent of extending that (not as far as I know).

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/BasePartitionProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/BasePartitionProcessor.cs]: I'm still a bit against marking it as `abstract`, at least for this preview. In case the user decides to use the factory approach to create partition processors, they would still need to create a useless type only to instantiate the generic `EventProcessor`, as its underlying type can't be abstract.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: The only `private` property I think we could set to `protected` is the `InnerClient`, in case we delegate the client creation to the `EventProcessor` in the future. We should keep it `private` if we keep things as they are.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: Just made a note about it. I'll open an issue once we're done with the PR.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: I'll open an issue when we're done with the PR.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs]: I'll keep it then. It won't get much worse than this (we might need to add an event hubs namespace here as well).

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/InMemoryPartitionManager.cs]: I'll do that once we're done with a PR. Just took a note.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Processor/InMemoryPartitionManagerTests.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Processor/InMemoryPartitionManagerTests.cs]: Clever. I have updated the lastModifiedTime we pass to the ownership so the test makes more sense.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Processor/EventProcessorLiveTests.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Processor/EventProcessorLiveTests.cs]: Agreed. I'll open (one more) issue. I'm glad we'll be getting a fourth preview.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs]: Added it to the delay. Partition Manager's methods do not support cancellation tokens. It should be simple to include it in this PR if you think we should.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: Sure. One more to the list.

Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: To tell you the truth, if this fails, it's an error on our part, not the user's fault. Not sure if it's worth discussing it.

This method is only called in two situations: upon a new ownership claim (right after adding it to ownership dictionary), and when we want to restart a failed pump (we filter the failed pumps from the dictionary's keys). If it throws, it means we called this method in a place we shouldn't, so we're simply throwing it and bringing everything down.

Does it make sense?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/Checkpoint.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/Checkpoint.cs]: There was discussion in the channel about that; allowing checkpoints to be created via the manager elsewhere, but it wasn't fully baked.   We can always adjust if needed.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: We could pass the whole ownership to the method instead of its partition id. Maybe this will make things more clear. It is guaranteed to exist.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: Better private and have to adjust than the other way around.   👍

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: Makes sense, I think.   If we've got tight control and guarantees about invariants, then I think we're good so long as we're (eventually) logging.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: A new overload seems like a good idea, indeed.

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs]: Partition Manager is user-supplied or one of our constructs?   If I recall, the normal use case is that it hangs out in partition context and is invoked by the user right?   I would think as an async method, it should have cancellation in there... but not a priority for preview 3.  Definitely something we want to get in for the actual blob storage thinger.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/BasePartitionProcessor.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/BasePartitionProcessor.cs]: I'm good with Initialize and Stop being just `virtual`, but I think that we need to force `ProcessEvents` and `ProcessError` to be implemented.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs]: nit:  I think you're missing "such" here....

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs]: It can be user-supplied, but we also provide two existing implementations (InMemory and BlobStorage).

It has three methods (all of them async):
 - ListOwnership: used by us for load balancing. I believe the user could make use of it as well.
 - ClaimOwnership: used by us for load balancing. I don't think it's useful for the user, as it requires a list of `PartitionOwnership`, a class with a `protected internal` constructor.
 - UpdateCheckpoint: indirectly called by the user via `PartitionContext`. We don't call it by ourselves. The user shouldn't call it directly, as it requires a `Checkpoint` instance, something they can't create.

I'm not sure if this gives you enough context to think about that, but we can discuss it for the next preview.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs]: Maybe my wording was not the best. I meant something like "This member is only used in case of failure, as other kinds of reasons will be specified by the event processor".

Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs]: Changed it to `This member is only used in case of failure. Shutdown and OwnershipLost close reasons will be specified by the event processor.`

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs]: Chose to using partitionId + sequenceNumber instead of passing the whole ownership.

File: sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample12_ConsumeEventsWithEventProcessor.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample12_ConsumeEventsWithEventProcessor.cs]: It is a partition processor's concrete type, but the generic event processor's underlying type.

Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample12_ConsumeEventsWithEventProcessor.cs]: Let me know if you still think it sounds weird.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs
Comment (kinelski) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionContext.cs]: I'll add some guards and let them throw in case it is `null`. This kind of situation can only happen when the user creates an `EventData` instance and tries to use it as `Checkpoint` instead of using an `EventData` received from the Event Hubs service.

This method can only be called from one of the `PartitionProcessor`'s methods (`Initialize`, `Close`, `ProcessEvents`, `ProcessError`) or constructor. Let the exception bubble up through them. We should already be handling these kinds of exceptions (or delegating them to our nonexistent error handler).

File: sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample12_ConsumeEventsWithEventProcessor.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample12_ConsumeEventsWithEventProcessor.cs]: That sounds smooth.   I like it!

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/tests/Infrastructure/EventProcessorManager.cs]: Yeah, I'd definitely say that, based on the context, we should make sure that we bubble cancellation down into it and respect it at that level.   As you said,  conversation for P4.    If you would be so kind, would you open an issue to track?

**Pull request merged**
